### ISSUE TITLE
Use own compiled mod in Docker server

### ISSUE THREAD

**Issue by hannesa2** [itzg/docker-minecraft-server#647]
Maybe it's a complete Minecraft beginner question (but I'm such a guy)
I followed https://github.com/itzg/docker-minecraft-server#using-the-data-volume to apply my own build mod to the docker server
* create my server `docker run -d -e VERSION=1.16.1 -v /home/hannes/mincraftData:/data -p 25565:25565 --restart unless-stopped --name minecraft -e EULA=TRUE itzg/minecraft-server:multiarch`
* using the artifact from [here](https://github.com/hannesa2/SuperTools/runs/1240887256) 

<img width="480" alt="image" src="https://user-images.githubusercontent.com/3314607/95719599-59f48680-0c70-11eb-8089-c49b28b18fc3.png">

* extract it
* copy `SuperTools-1.16.1-2-a4d6745.jar` to `/home/hannes/mincraftData/mods/`
* restart server `docker stop minecraft && docker start minecraft`

But my children don't see the mod at all.

Do you know what I'm missing, or how can I make it work in the server ?
Do the clients needs the mod as well ? I guess not, am I right ?

**Comment by itzg** [comment]
You're on the right track. Two things:

1. Since it's a Forge mod, you seem to be missing the  `-e TYPE=FORGE`
2. Regarding your question, for Forge mods, yes, the clients need the mods installed there too. What adds to the confusion is that some server types like Bukkit/Spigot/Paper, the **plugins** only need to be installed server-side, but have limited features compared to Forge mods.

**Comment by hannesa2** [comment]
Great, I followed your suggestion by

```
docker run -d -e TYPE=FORGE \
-v /home/hannes/mincraftData:/data \
-p 25565:25565 \
--restart unless-stopped \
--name minecraft \
-e FORGE_INSTALLER_URL=https://files.minecraftforge.net/maven/net/minecraftforge/forge/1.16.1-32.0.18/forge-1.16.1-32.0.18-installer.jar \
-e EULA=TRUE itzg/minecraft-server:multiarch
```

but in console `docker logs --follow minecraft` I see this 

>   Extra org/apache/commons/codec/net/QuotedPrintableCodec.class
>   Extra org/apache/commons/codec/net/RFC1522Codec.class
>   Extra org/apache/commons/codec/net/URLCodec.class
>   Extra org/apache/commons/codec/net/Utils.class
>   Data  io/netty/internal/
>   Processor failed, invalid outputs:
>     /data/./libraries/net/minecraft/server/1.16.1/server-1.16.1-slim.jar
>       Expected: 6cbe49c4d26497791d1d511c2d3b7d1fa46fa2e9
>       Actual:   3f1b64c684df16a25e6c44079b06ac12131db572
>     /data/./libraries/net/minecraft/server/1.16.1/server-1.16.1-extra.jar
>       Expected: 79ed2a48582a58ce917a04344c35c9c9abb557e2
>       Actual:   501b61ed1857f6452fe36a5faaf1265d38999c6f
> There was an error during installation
> 
> JVM info: Private Build - 1.8.0_265 - 25.265-b01
> java.net.preferIPv4Stack=true
> Found java version 1.8.0_265
> Target Directory: .
> Extractiung main jar:
>   Extracted successfully
> Considering minecraft server jar
> Downloading libraries
> Considering library net.minecraftforge:forge:1.16.1-32.0.18

I've no clue what's wrong with the hash.

**Comment by hannesa2** [comment]
Next try:
```
docker stop minecraft
docker rm minecraft
wget https://files.minecraftforge.net/maven/net/minecraftforge/forge/1.16.1-32.0.18/forge-1.16.1-32.0.18-installer.jar
```

<img width="694" alt="image" src="https://user-images.githubusercontent.com/3314607/96025310-38082900-0e55-11eb-84ce-58007a5f3f0c.png">

```
docker run -d -e TYPE=FORGE \
-v /home/hannes/mincraftData:/data \
-p 25565:25565 \
--restart unless-stopped \
--name minecraft \
-e FORGE_INSTALLER_URL=/home/hannes/forge-1.16.1-32.0.18-installer.jar \
-e EULA=TRUE itzg/minecraft-server:multiarch
```

but in console `docker logs --follow minecraft` I see this

```
[init] Running as uid=1000 gid=1000 with /data as 'drwxrwxr-x 8 1000 1000 4096 Oct 14 17:37 /data'
[init] Resolved version given LATEST into 1.16.3
[init] Resolving type given FORGE
[init] Downloading /home/hannes/forge-1.16.1-32.0.18-installer.jar ...
curl: (3) <url> malformed
[init] Failed to download from given location /home/hannes/forge-1.16.1-32.0.18-installer.jar
[init] Running as uid=1000 gid=1000 with /data as 'drwxrwxr-x 8 1000 1000 4096 Oct 14 17:37 /data'
[init] Resolved version given LATEST into 1.16.3
[init] Resolving type given FORGE
[init] Downloading /home/hannes/forge-1.16.1-32.0.18-installer.jar ...
curl: (3) <url> malformed
[init] Failed to download from given location /home/hannes/forge-1.16.1-32.0.18-installer.jar
[init] Running as uid=1000 gid=1000 with /data as 'drwxrwxr-x 8 1000 1000 4096 Oct 14 17:37 /data'
[init] Resolved version given LATEST into 1.16.3
[init] Resolving type given FORGE
[init] Downloading /home/hannes/forge-1.16.1-32.0.18-installer.jar ...
curl: (3) <url> malformed
[init] Failed to download from given location /home/hannes/forge-1.16.1-32.0.18-installer.jar
[init] Running as uid=1000 gid=1000 with /data as 'drwxrwxr-x 8 1000 1000 4096 Oct 14 17:37 /data'
[init] Resolved version given LATEST into 1.16.3
[init] Resolving type given FORGE
[init] Downloading /home/hannes/forge-1.16.1-32.0.18-installer.jar ...
curl: (3) <url> malformed
[init] Failed to download from given location /home/hannes/forge-1.16.1-32.0.18-installer.jar
```

**Comment by itzg** [comment]
I would not recommend not using the `FORGE_INSTALLER_URL`...in fact, I don't even remember why that option was added in the first place :)

If you're needing a specific Minecraft version along with Forge, `VERSION` can be used, such as

```
docker run -d -e TYPE=FORGE \
-v /home/hannes/mincraftData:/data \
-p 25565:25565 \
--restart unless-stopped \
--name minecraft \
-e VERSION=1.16.1 \
-e EULA=TRUE itzg/minecraft-server:multiarch
```

**Comment by itzg** [comment]
I should clarify: you don't need to use the installer variable since the container script will take care of downloading the forge installer automatically for the version requested.

**Comment by hannesa2** [comment]
```
docker run -d -e TYPE=FORGE \
-v /home/hannes/mincraftData:/data \
-p 25565:25565 \
--restart unless-stopped \
--name minecraft \
-e VERSION=1.16.1 \
-e EULA=TRUE itzg/minecraft-server:multiarch
```

Thnak you, it is much better now !
What I've observed the start is very "verbose". The output is **huge** with `docker logs --follow minecraft`
 
<img width="491" alt="image" src="https://user-images.githubusercontent.com/3314607/96084370-40954980-0ebf-11eb-8595-c76fa9c756ca.png">

### A
Is there a way to get rid of these verbose lines ? There are thousands  

### B
And I see this warnings, I've no clue if I can ignore it
```
[06:18:52] [Server thread/WARN]: Configuration file ./world/serverconfig/forge-server.toml is not correct. Correcting
[06:18:52] [Server thread/WARN]: Incorrect key server was corrected from null to SimpleCommentedConfig:{}
[06:18:52] [Server thread/WARN]: Incorrect key server.removeErroringEntities was corrected from null to false
[06:18:52] [Server thread/WARN]: Incorrect key server.removeErroringTileEntities was corrected from null to false
[06:18:52] [Server thread/WARN]: Incorrect key server.fullBoundingBoxLadders was corrected from null to false
[06:18:52] [Server thread/WARN]: Incorrect key server.zombieBaseSummonChance was corrected from null to 0.1
[06:18:52] [Server thread/WARN]: Incorrect key server.zombieBabyChance was corrected from null to 0.05
[06:18:52] [Server thread/WARN]: Incorrect key server.logCascadingWorldGeneration was corrected from null to true
[06:18:52] [Server thread/WARN]: Incorrect key server.fixVanillaCascading was corrected from null to false
[06:18:52] [Server thread/WARN]: Incorrect key server.dimensionUnloadQueueDelay was corrected from null to 0
[06:18:52] [Server thread/WARN]: Incorrect key server.clumpingThreshold was corrected from null to 64
[06:18:52] [Server thread/WARN]: Incorrect key server.treatEmptyTagsAsAir was corrected from null to false
[06:18:52] [Server thread/WARN]: Incorrect key server.fixAdvancementLoading was corrected from null to true
```

### C
I see this error, and I've no clue what's the root cause. Is it my SuperTools mod ?
```
[06:18:34] [main/INFO]: ModLauncher 6.1.1+74+master.966c698 starting: java version 1.8.0_265 by Private Build
[06:18:34] [main/INFO]: Added Lets Encrypt root certificates as additional trust
[06:18:34] [main/INFO]: SpongePowered MIXIN Subsystem Version=0.8 Source=file:/data/libraries/org/spongepowered/mixin/0.8/mixin-0.8.jar Service=ModLauncher Env=SERVER
[06:18:35] [main/INFO]: Launching target 'fmlserver' with arguments [--gameDir, .]
[06:18:44] [modloading-worker-1/INFO]: Forge mod loading, version 32.0.108, for MC 1.16.1 with MCP 20200625.160719
[06:18:44] [modloading-worker-1/INFO]: MinecraftForge v32.0.108 Initialized
[06:18:45] [modloading-worker-3/WARN]: Configuration file /data/config/supertools.toml is not correct. Correcting
[06:18:45] [modloading-worker-3/WARN]: Incorrect key SuperTools.Tool Attack Damage.Attack Damage.attackdamage_redstone was corrected from -0.5 to 0.0
[06:18:46] [Forge Version Check/INFO]: [forge] Starting version check at https://files.minecraftforge.net/maven/net/minecraftforge/forge/promotions_slim.json
[06:18:47] [Forge Version Check/INFO]: [forge] Found status: BETA Current: 32.0.108 Target: 32.0.108
[06:18:47] [Forge Version Check/INFO]: [supertools] Starting version check at https://raw.githubusercontent.com/TheKiritoPlayer20/SuperTools/updates/update.json
[06:18:47] [Forge Version Check/INFO]: [supertools] Found status: AHEAD Current: 1.1 Target: null
[06:18:48] [main/INFO]: Environment: authHost='https://authserver.mojang.com', accountsHost='https://api.mojang.com', sessionHost='https://sessionserver.mojang.com', name='PROD'
[06:18:49] [main/WARN]: Ambiguity between arguments [teleport, destination] and [teleport, targets] with inputs: [Player, 0123, @e, dd12be42-52a9-4a91-a8a1-11c01849e498]
[06:18:49] [main/WARN]: Ambiguity between arguments [teleport, location] and [teleport, destination] with inputs: [0.1 -0.5 .9, 0 0 0]
[06:18:49] [main/WARN]: Ambiguity between arguments [teleport, location] and [teleport, targets] with inputs: [0.1 -0.5 .9, 0 0 0]
[06:18:49] [main/WARN]: Ambiguity between arguments [teleport, targets] and [teleport, destination] with inputs: [Player, 0123, dd12be42-52a9-4a91-a8a1-11c01849e498]
[06:18:49] [main/WARN]: Ambiguity between arguments [teleport, targets, location] and [teleport, targets, destination] with inputs: [0.1 -0.5 .9, 0 0 0]
[06:18:49] [main/INFO]: Reloading ResourceManager: Default, forge-1.16.1-32.0.108-universal.jar, SuperTools-1.1.jar
[06:18:50] [Worker-Main-13/ERROR]: Parsing error loading recipe supertools:platinum_sickle
com.google.gson.JsonSyntaxException: Unknown item 'simplyplatinum:platinum_ingot'
	at net.minecraft.item.crafting.Ingredient.lambda$deserializeItemList$11(Ingredient.java:216) ~[?:?]
	at java.util.Optional.orElseThrow(Optional.java:290) ~[?:1.8.0_265]
	at net.minecraft.item.crafting.Ingredient.func_199803_a(Ingredient.java:215) ~[?:?]
	at net.minecraftforge.common.crafting.VanillaIngredientSerializer.parse(VanillaIngredientSerializer.java:41) ~[?:?]
	at net.minecraftforge.common.crafting.CraftingHelper.getIngredient(CraftingHelper.java:155) ~[?:?]
	at net.minecraft.item.crafting.Ingredient.func_199802_a(Ingredient.java:189) ~[?:?]
	at net.minecraft.item.crafting.ShapedRecipe.func_192408_a(ShapedRecipe.java:252) ~[?:?]
	at net.minecraft.item.crafting.ShapedRecipe.access$000(ShapedRecipe.java:24) ~[?:?]
	at net.minecraft.item.crafting.ShapedRecipe$Serializer.func_199425_a_(ShapedRecipe.java:276) ~[?:?]
	at net.minecraft.item.crafting.ShapedRecipe$Serializer.func_199425_a_(ShapedRecipe.java:272) ~[?:?]
	at net.minecraft.item.crafting.RecipeManager.func_215377_a(RecipeManager.java:141) ~[?:?]
	at net.minecraft.item.crafting.RecipeManager.func_212853_a_(RecipeManager.java:61) ~[?:?]
	at net.minecraft.item.crafting.RecipeManager.func_212853_a_(RecipeManager.java:38) ~[?:?]
	at net.minecraft.client.resources.ReloadListener.func_215269_a(SourceFile:13) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:670) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:646) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:456) ~[?:1.8.0_265]
	at net.minecraft.resources.AsyncReloader.func_219557_a(SourceFile:71) ~[?:?]
	at net.minecraft.resources.AsyncReloader.func_219560_a(SourceFile:70) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniCompletion.claim(CompletableFuture.java:543) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:667) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:646) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:575) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:649) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:456) ~[?:1.8.0_265]
	at net.minecraft.resources.AsyncReloader.func_219557_a(SourceFile:71) ~[?:?]
	at net.minecraft.resources.AsyncReloader.func_219560_a(SourceFile:70) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniCompletion.claim(CompletableFuture.java:543) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:667) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:646) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:575) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:649) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:456) ~[?:1.8.0_265]
	at net.minecraft.resources.AsyncReloader.func_219557_a(SourceFile:71) ~[?:?]
	at net.minecraft.resources.AsyncReloader.func_219560_a(SourceFile:70) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniCompletion.claim(CompletableFuture.java:543) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:667) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:646) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_265]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975) ~[?:1.8.0_265]
	at net.minecraft.resources.AsyncReloader$1.func_216873_a(SourceFile:53) ~[?:?]
	at net.minecraft.resources.AsyncReloader$1.func_216872_a(SourceFile:50) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:966) [?:1.8.0_265]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:940) [?:1.8.0_265]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) [?:1.8.0_265]
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1609) [?:1.8.0_265]
	at java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1596) [?:1.8.0_265]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) [?:1.8.0_265]
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) [?:1.8.0_265]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) [?:1.8.0_265]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175) [?:1.8.0_265]
```

**Comment by itzg** [comment]
A: possibly, but we're all basically at the mercy of forge and the mod authors with these

B: I assume most people ignore all these logs

C: the message has "supertools" in it, so yes, seems to be that mod causing it
